PixelMaze/PixelMaze.js

```diff
@@ -0,0 +1,182 @@
+      image: "tileset",
+      frame_size: Point.direction(TILE_SIZE, TILE_SIZE),
+    },
+    character: {
+      image: "character",
+      frame_size: Point.point(TILE_SIZE, 2 * TILE_SIZE),
+    },
+  },
+  sprites: {
+    walk: {
+      type: "directional",
+      spritesheet: "character",
+      start_row: 0,
+      frame_count: 4,
+      origin: ORIGIN_CHARACTER,
+    },
+    idle: {
+      type: "directional",
+      spritesheet: "character",
+      start_row: 4,
+      frame_count: 2,
+      origin: ORIGIN_CHARACTER,
+    },
+  },
+};
+
+const ARROW_KEYS = ["ArrowLeft", "ArrowRight", "ArrowUp", "ArrowDown"];
+
+class DPad {
+  constructor() {
+    this.direction = undefined;
+
+    this.keys_down = [false, false, false, false];
+  }
+
+  pressed(key) {
+    const direction = DPad.get_direction(key);
+    this.keys_down[direction] = true;
+    this.direction = direction;
+  }
+
+  released(key) {
+    const direction = DPad.get_direction(key);
+    this.keys_down[direction] = false;
+
+    // If there were other arrow keys pressed, switch to the first one
+    // we find.
+    const index = this.keys_down.findIndex((x) => x === true);
+    if (index !== -1) {
+      this.direction = index;
+    } else {
+      this.direction = undefined;
+    }
+  }
+
+  static get_direction(key) {
+    switch (key) {
+      case "ArrowLeft":
+        return GridDirection.LEFT;
+      case "ArrowRight":
+        return GridDirection.RIGHT;
+      case "ArrowUp":
+        return GridDirection.UP;
+      case "ArrowDown":
+        return GridDirection.DOWN;
+    }
+    return undefined;
+  }
+}
+
+const DPAD = new DPad();
+
+export const sketch = (p) => {
+  let canvas;
+
+  // p5.js specific resources
+  const p5_resources = {
+    images: {},
+  };
+
+  // logical resources that do not depend on p5.js
+  const resources = {
+    image_frames: {},
+    sprites: {},
+  };
+
+  let current_sprite;
+  let tilemap;
+  let player;
+
+  p.preload = () => {
+    preload_p5_resources(p, RESOURCE_MANIFEST, p5_resources);
+  };
+
+  p.setup = () => {
+    canvas = p.createCanvas(WIDTH, HEIGHT).elt;
+
+    parse_resources(RESOURCE_MANIFEST, p5_resources, resources);
+
+    current_sprite = new P5Sprite(
+      p5_resources.images.character,
+      resources.sprites.walk[GridDirection.LEFT]
+    );
+    tilemap = new P5Tilemap(
+      p5_resources.images.tileset,
+      new Tilemap(resources.image_frames.tileset_basic, INDICES)
+    );
+
+    player = new Player(
+      resources.sprites.walk,
+      resources.sprites.idle,
+      Point.direction(1, 6).scale(TILE_SIZE)
+    );
+
+    p.noSmooth();
+  };
+
+  p.draw = () => {
+    p.background(0);
+
+    p.push();
+    p.scale(TILE_SCALE, TILE_SCALE);
+    blit_tilemap(p, tilemap, Point.ORIGIN);
+
+    //const t = p.frameCount / 16.0;
+
+    player.update(p.frameCount);
+    const { position, sprite, t } = player.draw(p.frameCount);
+    blit_sprite(
+      p,
+      new P5Sprite(p5_resources.images.character, sprite),
```

Comment:
I need to think about how this should look in the long run. I prefer the decoupling, but at some point you need to map back the logical sprites to the actual image data...

Maybe all the resources store `id`s of things they reference? Then it's just a matter of doing a lookup into the p5 resources when done.